tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,204 @@
+    # we need the 'self' argument so method needs to be False, also it doesn't matter if signature faithful or not
+    args = CppSignatureGroup.from_native_function(f, method=False).most_faithful_signature().arguments()
+    code_list = [f"c10::IValue {args[i].name} = std::move(peek(stack, {i}, {len(args)}));" for i in
+                 range(len(args))] + [""]
+    binding_list = []
+    for i, arg in enumerate(args):
+        # expecting only Argument
+        if not isinstance(arg.argument, Argument):
+            raise Exception(f"Unexpected argument type, expecting `Argument` but got {arg}")
+        argument: Argument = arg.argument
+        unboxed_name, _, code = argumenttype_ivalue_convert(
+            argument.type, argument.name, mutable=argument.is_write
+        )
+        code_list.extend(code)
+        binding_list.append(arg.with_name(unboxed_name))
+    return binding_list, code_list
+
+
+# Takes in the type, name and mutability corresponding to an argument, and generates a tuple of:
+# (1) the C++ code necessary to unbox the argument
+# (2) A Binding corresponding to the newly created unboxed variable, including variable name and its CType
+def argumenttype_ivalue_convert(t: Type, arg_name: str, *, mutable: bool = False) -> Tuple[str, CType, List[str]]:
+    ctype = cpp.argumenttype_type(t=t, mutable=mutable, binds=arg_name).type
+
+    if isinstance(t, BaseType):
+        out_name = f"{arg_name}_base"
+        code = _gen_code_base_type(arg_name=arg_name, out_name=out_name, ctype=ctype)
+    elif isinstance(t, OptionalType):
+        out_name = f"{arg_name}_opt_out"
+        code = _gen_code_optional_type(arg_name=arg_name, out_name=out_name, t=t, ctype=ctype)
+    elif isinstance(t, ListType):
+        out_name = f"{arg_name}_list_out"
+        code = _gen_code_list_type(arg_name=arg_name, out_name=out_name, t=t, ctype=ctype)
+    else:
+        raise Exception(f"Cannot handle type {t}. arg_name: {arg_name}")
+    return out_name, ctype, code
+
+
+def _gen_code_base_type(arg_name: str, out_name: str, ctype: CType) -> List[str]:
+    ctype_str = ctype.cpp_type(strip_ref=True)
+    return [f"{ctype_str} {out_name} = {arg_name}.to<{ctype_str}>();"]
+
+
+def _gen_code_optional_type(arg_name: str, out_name: str, t: OptionalType, ctype: CType) -> List[str]:
+    ctype_str = ctype.cpp_type(strip_ref=True)
+    in_name = f"{arg_name}_opt_in"
+    res_name, _, res_code = argumenttype_ivalue_convert(t.elem, in_name)
+    return f"""
+c10::optional<c10::IValue> {arg_name}_opt = {arg_name}.toOptional<c10::IValue>();
+{ctype_str} {out_name};
+if ({arg_name}_opt.has_value()) {{
+    const c10::IValue {in_name} = {arg_name}_opt.value();
+    {connector.join(res_code)}
+    {out_name} = {ctype_str}({res_name});
+}} else {{
+    {out_name} = {ctype_str}();
+}}
+        """.split("\n")
+
+
+def _gen_code_list_type(arg_name: str, out_name: str, t: ListType, ctype: CType) -> List[str]:
+    ctype_str = ctype.cpp_type(strip_ref=True)
```

Comment:
nit: in the spirit of "avoid making strings until the last possible second", I'd just put `ctype.cpp_type(strip_ref=True)` directly inside of the f-string